### PR TITLE
C51-383: Quick Navigation Functionality

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR, the Quick Navigation Functionality(instead of static HTML) has been implemented.

## New Features
1. Dynamic Loading of Quick Navigation Months list.
2. When clicked on a month, if the first record of that month is already loaded, scroll to that record.
     If Not, load activities from the starting of that month and show Load More button on the up 
     direction as well.
3. Load More button design has been changed from text to Up and Down arrows.

## How it looks
after

### Backend
`Getmonthswithactivities` API has been updated, and now every month returns the number of records in that specific month. This is will be used in frontend to get data from a specific month.

### Frontend
**activity-month-nav.directive**
1. In `activity-month-nav.directive.html` dynamic markup has been implemented.
2. In `activity-month-nav.directive.js`, data loading logic has been implemented.
    - `setStartingOffsets` method, sets the starting offset for each months first record. 
    - When a month is clicked, if the first record is present on the UI, browser scrolls into that records using `scrollIntoView`.
    - If first record is not present, then an event is emitted `civicase::month-nav::clicked`, with the starting offset of that month.

**activity-feed.directive**
1. New variable `activityStartingOffset` is introduced. It keeps track of the starting offset, while making a call to the activities API.
2. This directive listens to `civicase::month-nav::clicked` event, and updates the `activityStartingOffset`. 
3. `pageNum` variable has been converted to an Object, so that it can keep track of page number in both UP and DOWN direction.
4. `checkIfRecordsAvailableOnDirection` method is introduced, and it returns true if more records are available of the given direction.
5. `previousPage` method is introduced, to handle upwards Load More.
6. `buildActivitiesArray` appends newly fetched records either to the Top or Bottom direction, based on the direction of Load More button.
7. In the activity object, `cardId` object is introduced which acts as a identification used by `month-nav` to scroll into.
8. `civicaseActivityFeed.query` event now also sends if the 'overdue_first' option is true.
9. Newly introduced `setActivityAPIOptions` abstracts the different options for the Activity API call.
10. Unit tests has been added/updated for the newly added code.
11. `activity-details-affix.directive.spec.js` is created, where code is moved from `activity-feed.directive.spec.js`.
12. `ActivityFilter.php` has been refactored, and a new function `updateParams` has been created. This is because the Activity filter is attached to the Activity API call https://github.com/compucorp/uk.co.compucorp.civicase/blob/2d8799a37c718db79991ba872eba59f1ecf5f6e5/civicase.php#L72, but for the `Getmonthswithactivities` API, the Activities are fetched directly without using the API, so we need to apply the `case` filters separately. 
So `CRM_Civicase_ActivityFilter::updateParams($params);` is called from the `Getmonthswithactivities` API.

### CODE REVIEW

File: ang/test/mocks/data/month-nav.data.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+(function (angular, _) {
+  var module = angular.module('civicase.data');
+
+  module.service('monthNavMockData', ['datesMockData', function (dates) {
+    var futureMonth = moment().add(1, 'month');
+    var currentMonth = moment();
+    var pastMonth = moment().add(-1, 'month');
+
+    var monthNavMockData = [
```
[the remark this answers is not in the document]
Comment (deb1990) [reply ang/test/mocks/data/month-nav.data.js]: as discussed, skipping, as of now.

File: ang/test/civicase/activity/feed/directives/activity-month-nav.directive.spec.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            initController();
+            $scope.$emit('civicaseActivityFeed.query', {}, params, false, overdueFirst);
+            $scope.$digest();
+          });
+
+          it('shows the future months', function () {
+            expect($scope.groups[0]).toEqual(
+              {
+                groupName: 'future',
+                records: [
```
[the remark this answers is not in the document]
Comment (deb1990) [reply ang/test/civicase/activity/feed/directives/activity-month-nav.directive.spec.js]: Fixed.

File: ang/test/civicase/activity/feed/directives/activity-feed.directive.spec.js [hunk 3]
```diff
@@ -39,8 +24,12 @@
 
       describe('loadActivities', function () {
         beforeEach(function () {
-          crmApi.and.returnValue($q.resolve({acts: {}}));
+          activitiesInCurrentPage = [];
+          allActivities = [];
+
+          mockActivitiesAPICall();
           initController();
+
```
[the remark this answers is not in the document]
Comment (deb1990) [reply ang/test/civicase/activity/feed/directives/activity-feed.directive.spec.js]: Honestly, I tried to leave a empty line to create a segregation between logic. The following lines are something which happens after controller is initialised, 

```
$scope.filters.activitySet = CaseTypes.get()['1'].definition.activitySets[0].name;
$scope.filters.activity_type_id = '5';
```

thats why I have grouped them together. So
1. variable declaration.
2. mock api and init controller
3. steps needs to be done after init controller.

File: ang/civicase/activity/feed/directives/activity-feed.directive.js [hunk 4]
```diff
@@ -260,7 +316,9 @@
      *
      * @return {Promise}
      */
-    function loadActivities () {
+    function loadActivities (mode) {
+      var options = setActivityAPIOptions(mode);
+
```
[the remark this answers is not in the document]
Comment (deb1990) [reply ang/civicase/activity/feed/directives/activity-feed.directive.js]: Not required, removed.

File: ang/civicase/activity/feed/directives/activity-feed.directive.js [hunk 5]
```diff
@@ -218,30 +279,25 @@
...
-        }
+        buildActivitiesArray(mode, newActivities);
+
         $scope.activityGroups = groupActivities($scope.activities);
         $scope.totalCount = allActivities.length;
-        $scope.remaining = remaining > 0 ? remaining : 0;
-        if (!allActivities.length && !pageNum) {
-          $scope.remaining = false;
-        }
+
```
[the remark this answers is not in the document]
Comment (deb1990) [reply ang/civicase/activity/feed/directives/activity-feed.directive.js]: Not really. Removed.

**Comment by deb1990** [comment]
> Do you think that civicase::month-nav::clicked event should be renamed to civicase::month-nav::record-miss. As we only trigger this event if the record are not in the present UI list.
Please correct me if I am wrong.

Changed it to `civicase::month-nav::set-starting-offset`.

**Comment by deb1990** [comment]
> 4. Need more space between year and ending months
Fixed

**Comment by deb1990** [comment]
> Selecting future date. doesn't scroll to the right position
fixed.

**Comment by deb1990** [comment]
> 5. Shows weird `Showing all %1` weird text while loading the data

fixed

**Comment by deb1990** [comment]
> If we can make the right sidebar with quick nav sticky. That would be a great add on

Its out of scope. we have https://compucorp.atlassian.net/browse/C51-397 for this feature.

**Comment by deb1990** [comment]
> Selecting past date, after loading data, doesn't scroll to the first record of the month and opens the records with a glitch.

Fixed

**Comment by swastikpareek** [comment]
I just noticed some glitches while changing the months

still a glitch

Also, if we click on any activity, we again see the loading placeholder, which we shouldn't technically appear
details is broken

Please let me know if there is any confusion

**Comment by deb1990** [comment]
> I just noticed some glitches while changing the months

Fixed.

**Pull request merged**